File: Adapters/MoPub/ObjC/Interstitial/AvocarrotMoPubInterstitialCustomEvent.m
//
//  AvocarrotMoPubInterstitialCustomEvent.m
//  AvocarrotSDKAdaptersObjC
//
//  Created by Glispa GmbH on 24.11.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

#import "AvocarrotMoPubInterstitialCustomEvent.h"
#import <AvocarrotInterstitial/AvocarrotSDK+AVOInterstitial.h>

@interface AvocarrotMoPubInterstitialCustomEvent ()

@property (nonatomic, strong) AVOInterstitial *interstitial;

@end

@implementation AvocarrotMoPubInterstitialCustomEvent

- (void)requestInterstitialWithCustomEventInfo:(NSDictionary *)info {
    if (!info) {
        NSLog(@"AvocarrotMoPubInterstitialCustomEvent - empty info!");
        [self requestDidFail];
        return;
    }
    
    NSString *adUnitId = info[@"adUnit"];
    if (!adUnitId) {
        NSLog(@"AvocarrotMoPubInterstitialCustomEvent - invalid adUnitId!");
        [self requestDidFail];
        return;
    }
    
    __weak AvocarrotMoPubInterstitialCustomEvent *weakSelf = self;
    
    [AvocarrotSDK.sharedSDK loadInterstitialWithAdUnitId:adUnitId
                                                 success:^(AVOInterstitial * _Nonnull interstitial) {
                                                     weakSelf.interstitial = [weakSelf preparedInterstitialWithCallbacksWithInterstitial:interstitial];
                                                     if (!weakSelf.interstitial) {
                                                         [weakSelf requestDidFail];
                                                     } else {
                                                         [weakSelf.delegate interstitialCustomEvent:weakSelf didLoadAd:weakSelf.interstitial];
                                                     }
                                                 } failure:^(AVOError * _Nonnull error) {
                                                     [weakSelf.delegate interstitialCustomEvent:weakSelf didFailToLoadAdWithError:error];
                                                 }];
}


- (void)showInterstitialFromRootViewController:(UIViewController *)rootViewController {
    [self.interstitial showFromViewController:rootViewController];
}


- (AVOInterstitial *)preparedInterstitialWithCallbacksWithInterstitial:(AVOInterstitial *)interstitial {
    if (!interstitial) {
        return nil;
    }
    
    __weak AvocarrotMoPubInterstitialCustomEvent *weakSelf = self;
    
    [[[[[interstitial onWillShow:^{
        [weakSelf.delegate interstitialCustomEventWillAppear:weakSelf];
    }] onDidShow:^{
        [weakSelf.delegate interstitialCustomEventDidAppear:weakSelf];
    }] onWillHide:^{
        [weakSelf.delegate interstitialCustomEventWillDisappear:weakSelf];
    }] onDidHide:^{
        [weakSelf.delegate interstitialCustomEventDidDisappear:weakSelf];
    }] onClick:^{
        [weakSelf.delegate interstitialCustomEventDidReceiveTapEvent:weakSelf];
    }];
    
    return interstitial;
}


- (void)requestDidFail {
    NSError *error = [[NSError alloc] initWithDomain:@"com.mopub.InterstitialCustomEvent" code:901 userInfo:nil];
    [self.delegate interstitialCustomEvent:self didFailToLoadAdWithError:error];
}

@end

File: Adapters/MoPub/Swift/AvocarrotMoPubInterstitialCustomEvent.swift
//
//  AvocarrotMoPubInterstitialCustomEvent.swift
//  AvocarrotSDKAdapters
//
//  Created by Glispa GmbH on 03.07.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

import UIKit
import MoPub
import AvocarrotInterstitial

@objc(AvocarrotMoPubInterstitialCustomEvent)
class AvocarrotMoPubInterstitialCustomEvent: MPInterstitialCustomEvent {
    private var interstitial : AVOInterstitial?

    override func requestInterstitial(withCustomEventInfo info: [AnyHashable : Any]?) {
        guard let customerInfo = info,
              let adUnitId = customerInfo["adUnit"] as? String else {
            print("AvocarrotMoPubInterstitialCustomEvent - invalid adUnitId")
            requestDidFail()
            return
        }

        AvocarrotSDK.shared.loadInterstitial(withAdUnitId: adUnitId,
                                             success: { [unowned self] (interstitial) in
                                                self.interstitial = self.preparedInterstitialWithCallbacksFor(interstitial: interstitial)
                                                guard let bnr = self.interstitial else {
                                                    self.requestDidFail()
                                                    return
                                                }
                                                self.delegate.interstitialCustomEvent(self, didLoadAd: bnr)
        }) { (error) in
            self.delegate.interstitialCustomEvent(self, didFailToLoadAdWithError: error)
        }
    }


    override func showInterstitial(fromRootViewController rootViewController: UIViewController?) {
        guard let vc = rootViewController,
              let fullscreen = interstitial else {
            return
        }

        fullscreen.show(from: vc)
    }


    private func preparedInterstitialWithCallbacksFor(interstitial: AVOInterstitial?) -> AVOInterstitial? {
        guard let fullscreen = interstitial else {
            return nil
        }

        fullscreen.onWillShow { [unowned self] in
            self.delegate.interstitialCustomEventWillAppear(self)
            }.onDidShow { [unowned self] in
                self.delegate.interstitialCustomEventDidAppear(self)
            }.onWillHide { [unowned self] in
                self.delegate.interstitialCustomEventWillDisappear(self)
            }.onDidHide { [unowned self] in
                self.delegate.interstitialCustomEventDidDisappear(self)
            }.onClick { [unowned self] in
                self.delegate.interstitialCustomEventDidReceiveTap(self)
        }

        return fullscreen
    }

    
    private func requestDidFail() {
        let error = NSError(domain: "com.mopub.InterstitialCustomEvent", code: 901)
        if let delegate = self.delegate {
            delegate.interstitialCustomEvent(self, didFailToLoadAdWithError: error)
        }
    }
}
